### PR TITLE
Resolve retain names issue

### PR DESCRIPTION
**Pull request by eliotwrobson**
Add retain names functionality to cross product. Resolves an inconsistency with expected behavior and should improve efficiency of minify function.

EDIT: Don't merge yet, just found a really dumb subtle bug in here. This also needs to wait until the shuffle product PR is merged, since that issue is present there as well (not a breaking bug).

### CODE REVIEW

**Comment by caleb531** [comment]
@eliotwrobson I just merged the shuffle PR, so you can resume development on this one. Just switch it from Draft to Open when you're ready for me to have a look at it.

File: automata/base/utils.py [hunk 1]
```diff
@@ -23,6 +23,26 @@ def freeze_value(value):
...
+    A helper function that returns a renaming function to be used in the creation of
+    other automata. The parameter counter should be an itertools count.
+    This helper function will return the same distinct output taken from counter
+    for each distinct input.
+    """
+
+    new_state_name_dict = {}
+
+    def renaming_function(item):
+        nonlocal new_state_name_dict
```
Comment (caleb531) [inline automata/base/utils.py:37]: @eliotwrobson You're not reassigning `new_state_name_dict` within this function, so the `nonlocal` declaration should be unnecessary.

**Comment by eliotwrobson** [comment]
@caleb531 should be good to go now. Found a bug earlier where the counters were being called too eagerly and the numbers were going higher than needed. So I added the helper function in the utils file that handles all that logic when a counter is provided.

This actually raises the question of whether it's worth it to rename states in the NFA operations, but those don't really have huge implications for efficiency. Because the DFA `minify` function has to make copies of state names, keeping the complexity of the state names down when that function is called is actually pretty important.

**Comment by eliotwrobson** [comment]
@caleb531 Oh no, I just meant that it might be worth adding `retain_names` logic to NFA constructions, but that it's not as big of a deal since the NFA operations aren't as sensitive to performance issues.

**Comment by caleb531** [comment]
@eliotwrobson Well, I suppose that _could_ make sense from a consistency standpoint, since DFA operations already support a `retain_names` parameter. However, if we want to go to the full mile, we would really need a `NFA.minify()` method (and associated per-operation `minify` parameters) as well.

**Comment by eliotwrobson** [comment]
@caleb531 unfortunately there are no known efficient algorithms for the NFA minimization problem (see [here](https://en.wikipedia.org/wiki/NFA_minimization)). My comment was really just an observation, I think things are fine as-is.

**Comment by caleb531** [comment]
@eliotwrobson Gotcha. Well, this looks good to me! Merging now.

**Review by caleb531 — APPROVED** [review]

**Pull request merged**